File: claude/harness/steambench_harness/envs/phantom.py
"""Phantom — a deterministic SteamBench arcade *memory / occlusion vision* env.

Dodge falling blocks, but every so often the lights go out: on a fixed blink
cycle ALL blocks vanish from view for a few ticks, then reappear. The blocks
keep falling (and still kill you) while hidden — so to survive you must
*remember* where they were and extrapolate, the one capability the six
fully-observable vision games never test. Both the structured state and the
rendered frame expose blocks ONLY while visible, so a state agent faces the same
partial-observability problem the pixel agent does.

Integer dynamics + Mulberry32 spawns (one draw per spawn) → bit-identical and
fully replayable across Python and TypeScript.
"""

from __future__ import annotations

from typing import Union

from ..protocol import AchievementSpec, ActionSpace, Env, Observation, VerifyMode
from ..registry import register

# Shared with runtime/pixel_game.py PhantomGame — keep in lock-step (parity test).
W, H = 168, 120
PW, PH = 20, 8
BW, BH = 14, 14
PLAYER_Y = H - PH - 3   # 109
PSPEED = 7
FALL = 4
SPAWN_GAP = 26
BLINK_PERIOD = 14       # blocks are visible for VISIBLE_TICKS of every BLINK_PERIOD
VISIBLE_TICKS = 9
_ACTIONS = ["left", "stay", "right"]


@register
class Phantom(Env):
    env_id = "arcade/phantom"
    appid = 9000015
    name = "Phantom"
    description = (
        "Dodge the falling blocks — but the lights blink out and the blocks "
        "vanish for a few ticks at a time while still falling. You must remember "
        "where they were and where they're going. Memory under partial "
        "observability: the frontier the fully-visible games never test."
    )
    verify_mode = VerifyMode.REPLAY
    action_space = ActionSpace(_ACTIONS)
    achievements = [
        AchievementSpec("survive_50", "Blink", "Survive 50 ticks.", 0.6),
        AchievementSpec("survive_150", "Afterimage", "Survive 150 ticks.", 0.3),
        AchievementSpec("survive_300", "Sixth Sense", "Survive 300 ticks.", 0.12),
        AchievementSpec("survive_600", "Echolocation", "Survive 600 ticks.", 0.035),
        AchievementSpec("survive_1200", "Mind's Eye", "Survive 1200 ticks.", 0.006),
    ]
    _LADDER = [(50, "survive_50"), (150, "survive_150"), (300, "survive_300"),
               (600, "survive_600"), (1200, "survive_1200")]

    def __init__(self) -> None:
        super().__init__()
        self.px = (W - PW) // 2
        self.blocks: list[list[int]] = []   # [x, y]
        self.alive = True
        self._since_spawn = SPAWN_GAP

    def reset(self, seed: int = 0) -> Observation:
        self._begin(seed)
        self.px = (W - PW) // 2
        self.blocks = []
        self.alive = True
        self._since_spawn = SPAWN_GAP
        return self._observe(0.0)

    def _visible(self) -> bool:
        return (self.steps % BLINK_PERIOD) < VISIBLE_TICKS

    def step(self, action: Union[int, str]) -> Observation:
        if self.done:
            return self._observe(0.0)
        name = self.action_space.name(action)
        self.steps += 1

        if name == "left":
            self.px = max(0, self.px - PSPEED)
        elif name == "right":
            self.px = min(W - PW, self.px + PSPEED)

        self._since_spawn += FALL
        if self._since_spawn >= SPAWN_GAP:
            self._since_spawn = 0
            x = self.rng.randrange(W - BW + 1)
            self.blocks.append([x, -BH])
        for blk in self.blocks:
            blk[1] += FALL
        self.blocks = [b for b in self.blocks if b[1] < H]

        # collision uses the REAL blocks, visible or not
        hit = False
        for bx, by in self.blocks:
            if (bx < self.px + PW and bx + BW > self.px and
                    by < PLAYER_Y + PH and by + BH > PLAYER_Y):
                hit = True
                break

        if hit:
            self.alive = False
            self.done = True
            reward = -1.0
        else:
            self.score += 1
            reward = 1.0
            for thresh, aid in self._LADDER:
                if self.score >= thresh:
                    self._unlock(aid)
        return self._observe(reward)

    def legal_actions(self) -> list[str]:
        return list(_ACTIONS)

    def render(self) -> str:
        vis = "VIS" if self._visible() else "DARK"
        return (f"score={int(self.score)} px={self.px} blocks={len(self.blocks)} "
                f"{vis} {'ALIVE' if self.alive else 'DEAD'}")

    def _observe(self, reward: float) -> Observation:
        visible = self._visible()
        # blocks are exposed ONLY while visible — the agent must remember the rest
        shown = [{"x": bx, "y": by} for bx, by in self.blocks] if visible else []
        return Observation(
            step=self.steps,
            state={
                "paddle_x": self.px,
                "paddle_y": PLAYER_Y,
                "paddle_w": PW,
                "paddle_h": PH,
                "paddle_speed": PSPEED,
                "block_w": BW,
                "block_h": BH,
                "fall": FALL,
                "width": W,
                "height": H,
                "visible": visible,
                "blocks": shown,
                "score": int(self.score),
                "alive": self.alive,
            },
            text=self.render(),
            legal_actions=self.legal_actions(),
            score=float(self.score),
            done=self.done,
            reward=float(reward),
            info={"score": int(self.score), "alive": self.alive, "visible": visible},
        )

File: claude/web/src/lib/arcade/phantom.ts
/**
 * Phantom — a deterministic SteamBench arcade *memory / occlusion vision* env.
 *
 * TypeScript mirror of `harness/steambench_harness/envs/phantom.py` (and the
 * headless twin of the pixel runtime's PhantomGame). Dodge falling blocks that
 * blink out of view on a fixed cycle — they keep falling (and killing) while
 * hidden, so you must remember them.
 *
 * Determinism note: integer dynamics; one rng draw per spawn (the column). The
 * blink is a pure function of `steps` (`steps % 14 < 9` = visible). Collision
 * uses the REAL blocks; the observation exposes them only while visible — but
 * the SCORE depends only on (seed, actions) via the real blocks, so a recorded
 * trace replays identically in Python and TypeScript.
 */

import { ActionSpace, Env } from "./base";
import type { AchievementSpec, Action, Observation } from "./types";

const W = 168;
const H = 120;
const PW = 20;
const PH = 8;
const BW = 14;
const BH = 14;
const PLAYER_Y = H - PH - 3; // 109
const PSPEED = 7;
const FALL = 4;
const SPAWN_GAP = 26;
const BLINK_PERIOD = 14;
const VISIBLE_TICKS = 9;
const ACTIONS = ["left", "stay", "right"] as const;

export class Phantom extends Env {
  static env_id = "arcade/phantom";
  static appid = 9000015;
  static displayName = "Phantom";
  static description =
    "Dodge the falling blocks — but the lights blink out and the blocks vanish " +
    "for a few ticks at a time while still falling. You must remember where they " +
    "were and where they're going. Memory under partial observability.";
  static verify_mode = "replay" as const;
  static action_space = new ActionSpace([...ACTIONS]);
  static achievements: AchievementSpec[] = [
    { id: "survive_50", name: "Blink", description: "Survive 50 ticks.", rarity_hint: 0.6 },
    { id: "survive_150", name: "Afterimage", description: "Survive 150 ticks.", rarity_hint: 0.3 },
    { id: "survive_300", name: "Sixth Sense", description: "Survive 300 ticks.", rarity_hint: 0.12 },
    { id: "survive_600", name: "Echolocation", description: "Survive 600 ticks.", rarity_hint: 0.035 },
    { id: "survive_1200", name: "Mind's Eye", description: "Survive 1200 ticks.", rarity_hint: 0.006 },
  ];
  private static LADDER: [number, string][] = [
    [50, "survive_50"], [150, "survive_150"], [300, "survive_300"],
    [600, "survive_600"], [1200, "survive_1200"],
  ];

  px: number;
  blocks: [number, number][];
  alive: boolean;
  private sinceSpawn: number;

  constructor() {
    super();
    this.px = Math.trunc((W - PW) / 2);
    this.blocks = [];
    this.alive = true;
    this.sinceSpawn = SPAWN_GAP;
  }

  reset(seed = 0): Observation {
    this.begin(seed);
    this.px = Math.trunc((W - PW) / 2);
    this.blocks = [];
    this.alive = true;
    this.sinceSpawn = SPAWN_GAP;
    return this.observe(0.0);
  }

  private visible(): boolean {
    return this.steps % BLINK_PERIOD < VISIBLE_TICKS;
  }

  step(action: Action): Observation {
    if (this.done) return this.observe(0.0);
    const name = this.actionSpace.name(action);
    this.steps += 1;

    if (name === "left") this.px = Math.max(0, this.px - PSPEED);
    else if (name === "right") this.px = Math.min(W - PW, this.px + PSPEED);

    this.sinceSpawn += FALL;
    if (this.sinceSpawn >= SPAWN_GAP) {
      this.sinceSpawn = 0;
      const x = this.rng.randrange(W - BW + 1);
      this.blocks.push([x, -BH]);
    }
    for (const b of this.blocks) b[1] += FALL;
    this.blocks = this.blocks.filter((b) => b[1] < H);

    let hit = false;
    for (const [bx, by] of this.blocks) {
      if (bx < this.px + PW && bx + BW > this.px && by < PLAYER_Y + PH && by + BH > PLAYER_Y) {
        hit = true;
        break;
      }
    }

    let reward: number;
    if (hit) {
      this.alive = false;
      this.done = true;
      reward = -1.0;
    } else {
      this.score += 1;
      reward = 1.0;
      for (const [thresh, aid] of Phantom.LADDER) {
        if (this.score >= thresh) this.unlock(aid);
      }
    }
    return this.observe(reward);
  }

  legalActions(): string[] {
    return [...ACTIONS];
  }

  render(): string {
    return `score=${Math.trunc(this.score)} px=${this.px} blocks=${this.blocks.length} ${this.visible() ? "VIS" : "DARK"} ${this.alive ? "ALIVE" : "DEAD"}`;
  }

  private observe(reward: number): Observation {
    const visible = this.visible();
    const shown = visible ? this.blocks.map(([x, y]) => ({ x, y })) : [];
    return {
      step: this.steps,
      state: {
        paddle_x: this.px, paddle_y: PLAYER_Y, paddle_w: PW, paddle_h: PH, paddle_speed: PSPEED,
        block_w: BW, block_h: BH, fall: FALL, width: W, height: H,
        visible, blocks: shown, score: Math.trunc(this.score), alive: this.alive,
      },
      text: this.render(),
      frame: null,
      legal_actions: this.legalActions(),
      score: this.score,
      done: this.done,
      reward: reward,
      info: { score: Math.trunc(this.score), alive: this.alive, visible },
    };
  }
}
